controller/src/main/java/io/pravega/controller/server/ControllerService.java

```diff
@@ -158,6 +160,37 @@
                 }, executor);
     }
 
+    public CompletableFuture<AddSubscriberStatus> addSubscriber(String scope, String stream, final String subscriber) {
+        Preconditions.checkNotNull(scope, "scopeName is null");
+        Preconditions.checkNotNull(stream, "streamName is null");
+        Preconditions.checkNotNull(subscriber, "subscriber is null");
+        Timer timer = new Timer();
+        return streamMetadataTasks.addSubscriber(scope, stream, subscriber, null)
```

Comment:
do you still need to route it through stream metadata tasks? you can perhaps directly write using the store apis.

Reply:
I'd prefer encapsulating the various steps like checking for streamExists etc inside StreamMetadataTasks..